### ISSUE TITLE
Dev REPL package checksum verification failure

### ISSUE THREAD

**Issue by rth** [pyodide/pyodide#2609]
Currently, I'm getting a checksum verification failure when installing numpy in the [dev REPL](https://pyodide.org/en/latest/console.html)

The error is,
```
None of the “sha256” hashes in the integrity attribute match the content of the subresource.
```

I suspect this is because the dev deployment has a cache duration of 1 hour and we overwrite files, so sometimes `packages.json` and actual packages might get temporarily out of sync and we would get these errors. 

I'm not sure if there is much we can do about it:
 - As far as I can tell after checking a few commits, package build is mostly reproducible and sha256 of packages remain constant unless there are build-related changes that are not very frequent. 
 - We can reduce the cache duration but then we would pay more in bandwidth if anyone uses dev packages extensively (right now there is no breakout of usage per deployed version)
 - Otherwise, we could disable sha256 verification on dev deployment but it's also not ideal

So hopefully this issue will resolve itself in the near future, opened it for future reference.

cc @amitsaha

**Comment by amitsaha** [comment]
Interesting side-effect!

Does that mean the caching logic for `packages.json` and the packages themselves are different? I am not sure about how client side caching works as well.

**Comment by rth** [comment]
> Does that mean the caching logic for packages.json and the packages themselves are different?

No, the caching logic is the same, but there is no guarantee that the cache with expire and new files re-downloaded for all files at the same time by the CDN. It depends on when a file was last accessed for a given CDN edge node / geographical location + any custom cache management logic they might have. So essentially we have no idea when it would happen.

**Comment by rth** [comment]
Closing for now. We can re-open if it happens again in the future.

**Comment by jobovy** [comment]
This seems to be happening again. Currently, I can't install `micropip` in the [dev REPL](https://pyodide.org/en/latest/console.html), because I get the error
```
Failed to find a valid digest in the 'integrity' attribute for resource 'https://cdn.jsdelivr.net/pyodide/dev/full/micropip-0.1-py3-none-any.whl' with computed SHA-256 integrity 'jSu3HzxpL2yOmdzzmpXISHrnCKa1irRHqZr7tGHLa9Y='. The resource has been blocked.
```
I checked the sha256 of the wheel and it doesn't match the one in ``repodata.json``, so I think it's the same issue again.

**Comment by ryanking13** [comment]
I think it is becuase we just released 0.21.3 and the CDN hasn't cached all files.

**Comment by jobovy** [comment]
Yes, this was happening a few hours ago already for me, before the 0.21.3 release.

**Comment by ryanking13** [comment]
Hmm, I'm not sure what we should do for this. Maybe we should add an option to disable integrity check and use it for dev consoles? @rth @hoodmane

**Comment by hoodmane** [comment]
Sounds great. I also have this problem sometimes locally because I built a package manually without updating `repodata.json` and it's a nuisance to have to run `buildall` which might rebuild a lot of things that don't actually need to be rebuilt. So I would like it if in development builds there was an easy way to disable the integrity check.

**Comment by rth** [comment]
> Maybe we should add an option to disable integrity check 

+1 as well.

**Comment by rth** [comment]
I think it would be good to maybe enable it in the dev REPL by default?

**Comment by ryanking13** [comment]
> I think it would be good to maybe enable it in the dev REPL by default?

Sounds good to me. Probably it is time to use some template engines for console.html (https://github.com/pyodide/pyodide/pull/3004#discussion_r952235324).

**Comment by KyleXie** [comment]
I was wondering if it's okay if we add a query parameter to the package url, something like `https://cdn.jsdelivr.net/pyodide/dev/full/micropip-0.1-py3-none-any.whl?hash=hash_value`, so that whenever the package changes, the hash_value changes, and the url changes, so in the end the cached package file from CDN will no longer be loaded.

**Comment by rth** [comment]
Thanks for the suggestion @KyleXie !

You are right that adding a query param to the URL does invalidate the browser cache (at least when tested on Firefox), however it seems to be ignored by the CDN.

For instance, when doing,
```bash
pip install httpie
http GET https://cdn.jsdelivr.net/pyodide/dev/full/micropip-0.5.0-py3-none-any.whl\?q\=a
http GET https://cdn.jsdelivr.net/pyodide/dev/full/micropip-0.5.0-py3-none-any.whl\?q\=b
```
in both cases I'm getting the following headers,
```
X-Cache: HIT, HIT
X-Content-Type-Options: nosniff
X-Served-By: cache-fra-etou8220109-FRA, cache-par-lfpg1960096-PAR
alt-svc: h3=":443";ma=86400,h3-29=":443";ma=86400,h3-27=":443";ma=86400
cache-control: public, max-age=3600
etag: "9a50c607c5b85ec5f000911b2cc47c73"
last-modified: Fri, 03 Nov 2023 05:22:40 GMT
```
which as far as I understand means they were served from the CDN cache, and not re-fetched from origin. Currently we set 1hour cache duration, I'm also not willing to reduce it by much more, as we would incur more bandwidth cost.

Maybe the only solution is to disable checksum verification on dev deployment. But it could then still lead to an inconsistent state (if there is an incompatible change between commits). Overall the point is that the dev deployment should only be used for testing and not anything production-oriented.

**Comment by KyleXie** [comment]
Thanks for your reply @rth !

I see your point. That makes a lot of sense. But for some CDN suppliers, it's configurable whether the query parameters will be ignored, something like this:
![image](https://github.com/pyodide/pyodide/assets/2870241/386fb2dc-ecca-42be-b837-f6f2d37403f7)

Another potential solution I can think of would be making the hash value a part of the file name, so the URL of the package would be something like:
```
https://cdn.jsdelivr.net/pyodide/dev/full/micropip-0.5.0-py3-none-any-hash-value.whl
```
Like how they handle those js/css files in Vue.js, the file name turns to something like `/assets/index-59e0e9a4.js`.
Not sure if this could work out.

**Comment by rth** [comment]
> But for some CDN suppliers, it's configurable whether the query parameters will be ignored, something like this:
image

Thanks for your suggestions, @KyleXie  I appreciate it!
I know, but we are relying on [JsDelivr CDN](https://www.jsdelivr.com/network) does an has an extra caching layer and aggregates multiple CDNs. Serving non immutable artifacts is not a primary use case there.

> Another potential solution I can think of would be making the hash value a part of the file name

You are right that could be a solution. The problem with it is,
 - if we don't delete old versions with checksums, our storage is going to grow for each commit on main which is not ideal
 - if we do delete everything , people can end up hitting dev artifacts that no longer exists.
 
 Overall, I would be +1 to close this as "won't fix". A good solution would be to have a separate dev bucket maybe with no caching, or deploy each commit on a separate url, but then someone would need to pay for bandwidth (and I don't think it's worth it). 
 
This issue is occasionally annoying, but I don't feel it's worth spending time on it, while there are more important issues address. Currently existing workarounds are:
 - it's possible to starts a REPL from CI artifacts of any commit on main (though admitedly it's a bit less convient to get that URL)
 - we should do alpha releases more often so there is a fixed release closer to the current `main` branch.

**Comment by KyleXie** [comment]
Agreed. Thanks for replying @rth
